Feat/424 validate location code on frontend

mgaseta opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->

# Description

Added location code validation to all necessary places.
- Create A Class Seedlot form
- Collection Step form
- Ownership Step form
- Interim Step form
- Extraction and Storage Step form

Notable changes:
- Disable location code input, if no agency is selected
- Limited the digits on each location code input to 2
- Show the correct error message according to the issue
- Validate the value entered

You may notice some that steps need refactoring... (Non-related with the new changes) But this PR is too big already, we can refactor when working on validations for each step.

Fixes #424 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Manually tested.

## Checklist

<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have already been accepted and merged



---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
[Backend](https://nr-spar-472-backend.apps.silver.devops.gov.bc.ca/)
[Frontend](https://nr-spar-472-frontend.apps.silver.devops.gov.bc.ca/)
[Oracle-API](https://nr-spar-472-oracle-api.apps.silver.devops.gov.bc.ca/)

Once merged, code will be promoted and handed off to following workflow run.
[Main Merge Workflow](https://github.com/bcgov/nr-spar/actions/workflows/merge-main.yml)

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
[Backend](https://nr-spar-472-backend.apps.silver.devops.gov.bc.ca/)
[Frontend](https://nr-spar-472-frontend.apps.silver.devops.gov.bc.ca/)
[Oracle-API](https://nr-spar-472-oracle-api.apps.silver.devops.gov.bc.ca/)

Once merged, code will be promoted and handed off to following workflow run.
[Main Merge Workflow](https://github.com/bcgov/nr-spar/actions/workflows/merge-main.yml)

craigyu reviewed (changes requested):
Really nice work!  💪😎👍
All inputs are working with informative status for users.

I ignored the brain overload code smells, but we should change the operator as suggested.
![image](https://github.com/bcgov/nr-spar/assets/25162561/214effb1-131e-4f3e-bbf9-648bfb4ba09d)

![image](https://github.com/bcgov/nr-spar/assets/25162561/18226fcf-cde2-48f5-9b90-448c4f04e8bb)

What should we do with this situation, 3 is between 0-99, but we want 03, perhaps we should format it for user when the number is less than 10?

mgaseta commented:
@craigyu 

Fixed operator: https://github.com/bcgov/nr-spar/pull/472/commits/04114210e1ad8c55413c1a993cf7eb581ca416e5

For the single location code character, I added the zero to the left on the blur function, it looks good, but maybe we need some validation from @SLDonnelly and the UX team.

Fixed https://github.com/bcgov/nr-spar/pull/472/commits/ecaaaf5db49f8707c392bcaaf9f7e457d9f416c6 and https://github.com/bcgov/nr-spar/pull/472/commits/3c6432c5818df601b0c14f038ed1288869ebcf07

RMCampos commented on frontend/src/components/ApplicantInformationForm/index.tsx:
```diff
@@ -222,15 +299,22 @@ const ApplicantInformationForm = () => {
+              onBlur={() => validateLocationCode()}
               onWheel={(e: React.ChangeEvent<HTMLInputElement>) => e.target.blur()}
-              helperText="2-digit code that identifies the address of operated office or division"
+              helperText={locHelper}
```
question (non-blocking): can we have a better name for the helper text? 😆

mgaseta replied:
For sure, fixed here: https://github.com/bcgov/nr-spar/pull/472/commits/c23553932e0f7ff40a421664c76b28364790c306
Nice catch ;)

RMCampos commented on frontend/src/components/SeedlotRegistrationSteps/CollectionStep/index.tsx:
```diff
@@ -65,15 +78,35 @@ const CollectionStep = (
-        isInvalid = true;
+    if (name === fieldsConfig.collector.name) {
+      if (!value) {
+        newValidObj.isLocationCodeInvalid = false;
```
question (non-blocking): If no value, then`isLocationCodeInvalid` shouldn't be true?

mgaseta replied:
Actually, no, the value is for the agency field, so if the location code was invalid but the user clears the agency field, then the location code must be cleared also and remove the invalid status.

RMCampos commented on frontend/src/components/SeedlotRegistrationSteps/ExtractionAndStorageStep/definitions.ts:
```diff
@@ -19,3 +19,11 @@ export const initialValidationObj: FormValidation = {
+  [key: string]: {
+    forestClientNumber: string,
+    invalidLocationMessage: string,
+    locHelper: string
```
question (non-blocking): same here? A better name always help :)

mgaseta replied:
Fixed here: https://github.com/bcgov/nr-spar/pull/472/commits/c23553932e0f7ff40a421664c76b28364790c306

RMCampos commented on frontend/src/components/SeedlotRegistrationSteps/InterimStep/index.tsx:
```diff
@@ -61,17 +70,73 @@ const InterimStorage = (
+    // Adding this check to add an extra 0 on the left, for cases where
+    // the user types values between 0 and 9
+    if (isInRange && locationCode.length === 1) {
+      locationCode = `0${locationCode}`;
```
thought: here we could have `locationCode.toString().padStart(2, '0')`

mgaseta replied:
Fixed here: https://github.com/bcgov/nr-spar/pull/472/commits/de07f326fb598aabac77746d59fa15c85a44f1f1

RMCampos reviewed (approved):
Great job! 🥇 💯 🎉 

I left some non-blocking thoughts and questions. Feel free to merge ❤️

mgaseta commented on frontend/src/components/SeedlotRegistrationSteps/ExtractionAndStorageStep/index.tsx:
```diff
@@ -52,22 +61,96 @@ const ExtractionAndStorage = (
+    // Adding this check to add an extra 0 on the left, for cases where
+    // the user types values between 0 and 9
+    if (isInRange && locationCode.length === 1) {
+      locationCode = `0${locationCode}`;
```
Nice, I'll try that!

mgaseta replied:
Waaaay fancier :wine_glass: :woman_dancing: 
Fix in this commit: https://github.com/bcgov/nr-spar/pull/472/commits/de07f326fb598aabac77746d59fa15c85a44f1f1

mgaseta commented on frontend/src/components/SeedlotRegistrationSteps/ExtractionAndStorageStep/index.tsx:
```diff
@@ -52,22 +61,96 @@ const ExtractionAndStorage = (
-          isInvalid = true;
+      case 'extractoryAgency':
+        if (!value) {
+          newValidObj.isExtractorCodeInvalid = isInvalid;
```
It's the same case as the other one, when setting an empty value to the agency, clears the location code and removed invalid status.

craigyu reviewed (approved):
Approved! Waiting for @RMCampos to resolve the conversations or further comments then we can merge
